### ISSUE TITLE
apitrace replay crashes in __memcpy_avx_unaligned

### ISSUE THREAD

**Issue by kparal** [apitrace/apitrace#407]
If I record an XCOM game trace and try to replay that, the replay consistently crashes early in the process (at random places, usually in about 30-60 seconds). I have tested this with several different XCOM traces. However, I could not reproduce that with different (simple, non-steam) games like extremetuxracer or 0ad.

The crash looks like this:
```
$ apitrace replay game.x86_64.trace
apitrace: warning: caught signal 11
46593800: error: caught an unhandled exception
/usr/bin/glretrace+0x26f19c
/lib64/libpthread.so.0+0x109ef
/lib64/libc.so.6+0x148c06
/usr/lib64/dri/radeonsi_dri.so+0x35a8fe
/usr/lib64/dri/radeonsi_dri.so+0x35c70a
/usr/lib64/dri/radeonsi_dri.so+0x1dc8be
/usr/lib64/dri/radeonsi_dri.so+0x1ae827
/usr/lib64/dri/radeonsi_dri.so+0x1aeb33
/usr/bin/glretrace+0xec864
/usr/bin/glretrace+0x4ce36
/usr/bin/glretrace+0x4d727
/usr/bin/glretrace+0x4d991
/lib64/libpthread.so.0+0x7609
/lib64/libc.so.6: clone+0x6c
?
apitrace: info: taking default action for signal 11
```

When I use gdb to connect to the `glretrace` process and wait until it crashes, I see this:
```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ff73bfff700 (LWP 32420)]
__memcpy_avx_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:238
238     rep movsb
```

The full gdb backtrace is here:
[gdb.backtrace.txt](https://github.com/apitrace/apitrace/files/68921/gdb.backtrace.txt)

I'm not a developer, is this an apitrace bug, or maybe a compiler issue?

I'm using today's version of apitrace (83032cb) on Fedora 23, compiled using [Fedora's spec file](http://pkgs.fedoraproject.org/cgit/apitrace.git/tree/apitrace.spec?h=f23).
Installed packages:
```
apitrace-7.1.20151221.git83032cb-1.fc23.x86_64
kernel-4.2.8-300.fc23.x86_64
mesa-dri-drivers-11.0.6-1.20151122.fc23.x86_64
xorg-x11-drv-ati-7.6.0-0.4.20150729git5510cd6.fc23.x86_64
xorg-x11-server-Xorg-1.18.0-2.fc23.x86_64
llvm-libs-3.7.0-1.fc23.x86_64
gcc-c++-5.3.1-2.fc23.x86_64
libstdc++-5.3.1-2.fc23.x86_64
glibc-2.22-6.fc23.x86_64
```
Graphics:
```
01:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Curacao PRO [Radeon R7 370 / R9 270/370 OEM] [1002:6811]
OpenGL renderer string: Gallium 0.4 on AMD PITCAIRN (DRM 2.43.0, LLVM 3.7.0)
OpenGL core profile version string: 4.1 (Core Profile) Mesa 11.0.6 (git-2555e00)
OpenGL core profile shading language version string: 4.10
OpenGL version string: 3.0 Mesa 11.0.6 (git-2555e00)
OpenGL shading language version string: 1.30
```

**Comment by jrfonseca** [comment]
This could be a bug in glretrace (maybe it recorded/replayed a truncated blob), or a bug in the radeon_si driver (maybe it's assuming the blob is bigger than it actually should be.)

Could you provide me the trace (via dropbox/gdrive/etc), and which call nos it crashes the most?

**Comment by kparal** [comment]
The trace is here (it's 1.5GB compressed):
https://drive.google.com/file/d/0B0Opr_geiK5nWUMwVEhJVnZPR2s/view?usp=sharing

Please note that this trace is supposed to contain a gpu hang reproducer for the radeonsi driver:
https://bugs.freedesktop.org/show_bug.cgi?id=80419#c73
Eventually I managed to get a gpu hang when replaying this trace (see the next comment), so it seems it does. But I can reproduce the same glretrace crash even in other XCOM traces which didn't contain a gpu hang during the recording. So I'm not sure how closely these two issues are related. There seems to be something very fishy in there, because both seem to point to `__memcpy_avx_unaligned`.

I can record and upload a different trace file, if you prefer.

> which call nos it crashes the most?

This is my first time using apitrace, please excuse my dumbness. If you refer to those numbers next to the crash, e.g.:
```
46593800: error: caught an unhandled exception
```
then it crashes completely randomly. I have tried to run it a few times, and I received these numbers:
```
36098102
6713619
11560053
13368473
30363117
8865026
```
I don't think there are ever two same numbers. Also, after I reported this, I managed to get 2 full replays which didn't crash at all (in my ~10 attempts before reporting this, it always crashed). So this seems to be very racy.

**Comment by kparal** [comment]
Here's a newer gdb backtrace with mesa git ea8c0b1 (2012-12-21):
[gdb.backtrace2.txt](https://github.com/apitrace/apitrace/files/69636/gdb.backtrace2.txt)

gdb output was:
```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe7ba6700 (LWP 2743)]
__memcpy_avx_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:66
66      vmovdqu -0x20(%rcx), %xmm14
```

**Comment by jrfonseca** [comment]
Thanks.

I can't replay the provided trace on NVIDIA by some misterious reason:

    16714 @3 glXMakeCurrent(dpy = 0x51d3fc0, drawable = 52428819, ctx = 0x7fc4428bf700) = True
    error: failed to make current OpenGL context and drawable

But when I replay the trace on llvmpipe I get

    27018 @3 glDrawRangeElementsBaseVertex(mode = GL_TRIANGLES, start = 0, end = 215, count = 690, type = GL_UNSIGNED_SHORT, indices = NULL, basevertex = 0)
    warning: index out of range
    warning: index out of range
    warning: index out of range
    ...

That is, the application is giving a hint to the OpenGL driver that indices are between 0 and 215, but in fact in call 21933, the index at position 164 is 216, and more later.

I wouldn't be surprised that the segfaults with radeon_si driver are caused by this too.

This apitrace patch tweaks the recorded calls so that the application hints are ignored:

    diff --git a/retrace/glretrace.py b/retrace/glretrace.py
    index 5b491e8..a2e2f2a 100644
    --- a/retrace/glretrace.py
    +++ b/retrace/glretrace.py
    @@ -270,6 +270,10 @@ class GlRetracer(Retracer):
                 function.name == 'glBegin'
             )
     
    +        if function.name.startswith('glDrawRangeElements'):
    +            print r'    start = 0;'
    +            print r'    end = 0x7fffffff;'
    +
             # Keep track of active program for call lists
             if function.name in ('glUseProgram', 'glUseProgramObjectARB'):
                 print r'    if (currentContext) {'

Try this patch too, and see if your crashes go away.

If they do, you'll need to file a bug against Mesa / Radeon driver, so that the driver does this sort of workaround internally for this application.  Refer this bug.

**Comment by kparal** [comment]
I can confirm this patch fixes the glretrace crash. If I understand you correctly, this is a game bug, and should be fixed by game developers, right?

Thanks a lot for your work and quick feedback.

**Comment by jrfonseca** [comment]
> If I understand you correctly, this is a game bug, 

Yes.

> and should be fixed by game developers, right?

Ideally, yes.

But if this is not an open-source game, it might be worth to let the Mesa developers know, as they might be interested in working around the issue in the drivers. 

There are in fact [several bugs reported on FDO against XCOM](https://bugs.freedesktop.org/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&list_id=565274&query_format=advanced&short_desc=XCOM&short_desc_type=allwordssubstr).

**Comment by kparal** [comment]
Thanks. For future reference and anyone stumbling into this ticket, the radeonsi/mesa side of the issue is being dealt with in here:
https://bugs.freedesktop.org/show_bug.cgi?id=80419#c87

**Comment by kparal** [comment]
One more question, I wonder whether the patch is going to be pushed to master to workaround similar issues in the future, or whether I should maintain it locally until this is worked around in my driver? (will that also fix glretrace crashes?)

**Comment by jrfonseca** [comment]
> I wonder whether the patch is going to be pushed to master to workaround similar issues in the future

No. apitrace's purpuse is to help application developers to find bugs, not hide them under the carpet.

It might be a good idea is to add to apitrace logic to detect this bugs.  But I don't have the time I'm afraid.

**Comment by nhaehnle** [comment]
I think this bug report should be re-opened.

Plenty of bugs to go around in all involved applications. Apparently, there's one in XCOM, there's certainly one in radeonsi, but it seems there is also one in apitrace. Playing back the trace by Kamil in Valgrind with our driver, I get

```
70524 @3 glVertexAttribPointer(index = 0, size = 2, type = GL_SHORT, normalized = GL_FALSE, stride = 12, pointer = blob(88))
70525 @3 glVertexAttribPointer(index = 1, size = 4, type = GL_UNSIGNED_BYTE, normalized = GL_TRUE, stride = 12, pointer = blob(88))
70526 @3 glVertexAttribPointer(index = 2, size = 4, type = GL_UNSIGNED_BYTE, normalized = GL_TRUE, stride = 12, pointer = blob(88))
70527 @3 glDrawRangeElementsBaseVertex(mode = GL_TRIANGLES, start = 0, end = 15, count = 30, type = GL_UNSIGNED_SHORT, indices = blob(60), basevertex = 0)
==6319== Thread 4:
==6319== Invalid read of size 2
==6319==    at 0x4C3076E: memcpy@@GLIBC_2.14 (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==6319==    by 0x95C92EB: memcpy (string3.h:53)
==6319==    by 0x95C92EB: u_upload_data (u_upload_mgr.c:252)
==6319==    by 0x95CAF74: u_vbuf_upload_buffers (u_vbuf.c:990)
==6319==    by 0x95CAF74: u_vbuf_draw_vbo (u_vbuf.c:1264)
==6319==    by 0x94447A6: st_draw_vbo (st_draw.c:291)
...
==6319==  Address 0x180bea88 is 0 bytes after a block of size 88 alloc'd
==6319==    at 0x4C2C84F: operator new[](unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==6319==    by 0x65948E: Blob (trace_model.hpp:331)
==6319==    by 0x65948E: trace::Parser::parse_blob() (trace_parser.cpp:866)
==6319==    by 0x65A90E: parse_value (trace_parser.hpp:165)
==6319==    by 0x65A90E: trace::Parser::parse_arg(trace::Call*, trace::Parser::Mode) (trace_parser.cpp:616)
==6319==    by 0x65B00D: trace::Parser::parse_call_details(trace::Call*, trace::Parser::Mode) (trace_parser.cpp:497)
```

Observe that we're trying to upload vertex attribute data from a memory region that is 88 bytes long - corresponding to the calls to `glVertexAttribPointer`. Furthermore, observe that 88 bytes is way too small for 16 entries (end - start + 1) of 12 bytes stride.

Looking by hand at the index data supplied to `glDrawRangeElementsBaseVertex`, I do notice that only indices 0 to 7 are referenced, for which the 88 bytes are sufficient.

In summary, I'd say that apitrace is trying to be too clever, here, reducing the size of the blob that is stored. This is a mistake, because the OpenGL spec says "However, the implementation may partially process unused vertices, (...)" in the explanation of `DrawRangeElements` (section 10.4 of the OpenGL 4.5 spec).
